Feature: Allow automatically separating vehicles in shared orders

twpol opened the pull request:
This is my proposed code for allowing vehicles (of all types) to be automatically separated amongst their shared orders with the click of a single button (once for each shared order), to massively reduce the micromanagement needed for optimal passenger journeys.

I filed issue #8095 initially to check that the feature was acceptable, and it was suggested I go forward with a PR.

As this is my first-time contributing code to OpenTTD, even though I have read the contributing guidelines and linked materials, I will also probably have missed/misunderstood something so please let me know where and how I might have messed up.

As part of that, I have tried to document the algorithm itself and surrounding code with comments but if anything isn't clear, or the details would be better served elsewhere, let me know.

### Outstanding issues - https://github.com/OpenTTD/OpenTTD/pull/8342#issuecomment-1867818309

- [ ] When should automatic separation be unavailable?
- [x] Integration with timetables
- [ ] Queuing detection and behaviour

twpol committed:
Feature #8095: Allow automatically separating vehicles in shared orders

James103 commented on src/base_consist.h:
```diff
@@ -23,6 +23,9 @@ struct BaseConsist {
 	Date timetable_start;               ///< When the vehicle is supposed to start the timetable.
 
+	Date first_order_last_departure;    ///< When the vehicle last left the first order.
+	Ticks first_order_round_trip_time;   ///< How many ticks for a single circumnavigation of the orders.
```
For consistency
```suggestion
	Ticks first_order_round_trip_time;  ///< How many ticks for a single circumnavigation of the orders.
```

James103 commented:
Commit messages must follow a specified format. Instead of "Update ...", you should either merge that commit into your previous commit or use a commit message that's something like "Fix: grammar".

twpol commented:
> Commit messages must follow a specified format. Instead of "Update ...", you should either merge that commit into your previous commit or use a commit message that's something like "Fix: grammar".

Yes, that's Github's fault because I simply clicked to apply the suggested fix and it wasn't in that form. I fixed it locally and pushed an updated version now.

LordAro commented on src/saveload/saveload.h:
```diff
@@ -304,6 +304,8 @@ enum SaveLoadVersion : uint16 {
 	SLV_ENDING_YEAR,                        ///< 218  PR#7747 v1.10 Configurable ending year.
 	SLV_REMOVE_TOWN_CARGO_CACHE,            ///< 219  PR#8258 Remove town cargo acceptance and production caches.
 
+	SLV_AUTOMATIC_SEPARATION,               ///< 220
```
this can be filled in now

LordAro commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,87 @@ void Vehicle::HandleLoading(bool mode)
+		char buffer[128];
+		SetDParam(0, this->index);
+		GetString(buffer, STR_VEHICLE_NAME, lastof(buffer));
+		DEBUG(misc, 4, "Orders %p RTT = %d [%.2f days] [%d veh], gap = %d [%.2f days] [%d veh + %d q] / %s gap = %d [%.2f days], wait = %d [%.2f days]", this->orders, round_trip_time, (float)round_trip_time / DAY_TICKS, round_trip_count, separation, (float)separation / DAY_TICKS, vehicles_operating, vehicles_queuing, buffer, now - last_departure, (float)(now - last_departure) / DAY_TICKS, this->first_o ...
```
as per compile warnings. pointer value isn't particularly useful anyway, is there an id that can be used instead?

twpol replied:
I don't see anything particularly useful on the `OrderList` class, except the first vehicle (`Vehicle *first_shared`). It's also logging the vehicle that is departing, so maybe trying to identify the order list itself is not needed.

LordAro reviewed (commented):
Thanks for this! Code looks fine on brief inspection. Will require some actual testing though :)

andythenorth commented:
Testing this, I got an assert, but I do not know if that is related to the current mac build, or specific to this PR :) 

I've seen the assert twice in short succession.  I can't get it to trigger in current master using same savegame.

Crash log attached; the savegame will be pretty useless, it contains multiple unreleased grfs.

[Archive.zip](https://github.com/OpenTTD/OpenTTD/files/5675329/Archive.zip)

JGRennison commented:
> Testing this, I got an assert, but I do not know if that is related to the current mac build, or specific to this PR :)
> 
> I've seen the assert twice in short succession. I can't get it to trigger in current master using same savegame.
> 
> Crash log attached; the savegame will be pretty useless, it contains multiple unreleased grfs.

You'll probably want to locally cherry-pick https://github.com/OpenTTD/OpenTTD/commit/c98717cb4518d80ccccb2ae1677d0a394542f129 for testing on Apple/LLVM.

twpol commented:
> You'll probably want to locally cherry-pick [c98717c](https://github.com/OpenTTD/OpenTTD/commit/c98717cb4518d80ccccb2ae1677d0a394542f129) for testing on Apple/LLVM.

Would it be helpful for people testing if I rebased this PR onto that commit (or another newer commit)?

LordAro commented:
Always :)

twpol commented:
@andythenorth Hopefully the builds just generated will be more successful for you. 🙂

andythenorth commented:
No assert.  UI seems straightforward.  

Simple case works as expected for me in test cases with ships, trams, trains.  No full load orders, no explicit timetable, no conditional orders.  Works great.

Cases that were expected to 'break' the separation and do:

- using full load orders on some stations
- setting wait times at some stations

Cases I thought would break, but appear to work:

- 3 different tram routes, each busy, all including an order for the same 1 tile station; separation appears to be maintained well for all 3 routes
-  1 route severely overloaded with trams, with close stations so that tram jams can block station exits; the algorithm handled this surprisingly well; there were still jams, but trams mostly spaced out well

Cases I didn't try that I would expect to break:

- conditional orders

twpol commented:
I'm glad things went well; your results are mostly what I'd expect.

To expand a bit on the expectations from me:

* Full load is itself a separation method, so it's not compatible. I don't see a reasonable way to use both features together, and automatic separation was inspired by the difficulty in vehicle spacing in cases where full load is not appropriate. I'm open to making this explicit in the help text and/or UI, e.g., by disabling the automatic separation button if there are full load orders.
* An explicit timetable is also currently not compatible, but I wonder if we could reset the timetable lateness via automatic separation to make them work together?
* Conditional orders might be okay if there is a stable average journey time for the group of shared order vehicles (e.g., each vehicle always takes the same route). If that average is unstable (e.g., routes are based on demand), you'll have problems with automatic separation.

However, I'm particularly intrigued by this:

> * setting wait times at some stations

Are you referring to setting the stay time at a station in the timetable? In my testing, this seems to behave itself. It makes the vehicles wait for the timetabled time instead of their loading time, but that time is all included in the automatic separation measurement of journey time, so the automatic separation waits at the first station take account of that.

Where you might have problems, is if the wait time causes vehicles to queue up on the approach to the first order station. In this scenario, automatic separation believes itself to be responsible for that queue and reduces the wait times accordingly.

2TallTyler commented:
I agree about disabling the automatic separation button if there are full load orders. Is it possible to have a tooltip explaining why the button is greyed out?

ldpl commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,87 @@ void Vehicle::HandleLoading(bool mode)
+	 * vehicles, so that the extra time queuing does not have an adverse effect on separation */
+	int round_trip_time = round_trip_count > 0 ? round_trip_sum / round_trip_count : 0;
+	int vehicles = vehicles_operating + vehicles_queuing;
+	int separation = max(1, vehicles > 0 ? (int)(round_trip_time * ((float)vehicles_operating / vehicles) / (int)vehicles) : 1);
```
Shouldn't be using float as it's not desync-safe. Also wouldn't hurt to add a comment why exectly are queeing vehicles a problem and how is extra (vehicles_operating / vehicles) solving it.

twpol replied:
I've removed the float usage and refactored the code and written a big comment which, I hope, explains the algorithm and logic behind it. It's slightly altered from the version reviewed here, but play tested to be suitable and is easier to explain. 🙂

andythenorth commented:
> Where you might have problems, is if the wait time causes vehicles to queue up on the approach to the first order station. In this scenario, automatic separation believes itself to be responsible for that queue and reduces the wait times accordingly.

Yes this is exactly the case.  I set a wait at the first station in the order, and the station (roadstop) has capacity for 1 vehicle.  I don't have the savegame, and I can't remember the exact outcome, but I could try and repro if it helps.  However I had no expectation that case would be handled, and I don't think it needs to be.

TrueBrain commented:
Would you mind rebasing your Pull Request? It would allow us to create a preview (new feature). Tnx :)

TrueBrain commented on src/command.cpp:
```diff
@@ -94,6 +94,8 @@ CommandProc CmdModifyOrder;
 CommandProc CmdDeleteOrder;
 CommandProc CmdInsertOrder;
+CommandProc CmdOrderAutomaticSeparation;
+
```
Random extra newline :D Not sure if it was intended?

twpol replied:
Removed

TrueBrain commented on src/order_gui.cpp:
```diff
@@ -555,6 +555,16 @@ struct OrdersWindow : public Window {
+	 */
+	void OrderClick_AutomaticSeparation()
+	{
+		if (DoCommandP(this->vehicle->tile, this->vehicle->index, !this->vehicle->AutomaticSeparationIsEnabled(), CMD_ORDER_AUTOMATIC_SEPARATION | CMD_MSG(STR_ERROR_CAN_T_DELETE_THIS_ORDER))) {
```
The error message seems wrong for this command?

twpol replied:
Ah, I forgot this. But I am also not sure what error message, if any, should go here. The command can only fail if you don't own the vehicle or try and run it on a non-primary vehicle, neither of which I believe to be possible through the GUI.

For now, I've removed the error message here, but please do let me know how this should be handled if that's not right.

TrueBrain commented on src/saveload/saveload.h:
```diff
@@ -304,6 +304,8 @@ enum SaveLoadVersion : uint16 {
 	SLV_ENDING_YEAR,                        ///< 218  PR#7747 v1.10 Configurable ending year.
 	SLV_REMOVE_TOWN_CARGO_CACHE,            ///< 219  PR#8258 Remove town cargo acceptance and production caches.
 
```
I think this newline shouldn't be here :)

stormcone replied:
The versions are grouped by five, and this is the first piece in the next group. So I think it should be. :)

TrueBrain replied:
Valid point! Did not see that :D Cheers!

TrueBrain commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
+	 * to over-compensate rather than aim exactly for the ideal (which is very approximate here). */
+	int round_trip_time = round_trip_count > 0 ? round_trip_sum / round_trip_count : 0;
+	int vehicles_moving_ratio = max(1, vehicles - 2 * vehicles_queuing);
+	int separation = max(1, vehicles > 0 ? ((round_trip_time * vehicles_moving_ratio) / vehicles) / vehicles : 1);
```
`a * b / c / d` == `((a * b) / c) / d`. Adding those extra `()` doesn't add anything, but it suggests it does, confusing the reader it is something special. I suggest removing them :)

twpol replied:
It's been a while since I did C++ like this, thanks for reminded me; parentheses removed.

TrueBrain commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
+	this->first_order_last_departure = max(last_departure + separation, now);
+
+	/* Debug logging can be quite spammy as it prints a line every time a vehicle departs the first manual order */
+	if (_debug_misc_level >= 4) {
```
Who is going to be the user of this debug statement? I ask, as I wonder if it is useful to keep in now you did your initial work? Honest question, more of thinking out loud :)

twpol replied:
It's a good question. 🙂

Honestly, I'd like to keep the code, both for future development and curious/bug reporting users, but debug misc=4 might well be the wrong door to put it behind.

A higher debug misc level, or a compile-time constant (if OpenTTD uses those for debugging code)?

TrueBrain commented on src/vehicle_base.h:
```diff
@@ -771,6 +771,11 @@ struct Vehicle : VehiclePool::PoolItem<&_vehicle_pool>, BaseVehicle, BaseConsist
 	inline void SetServiceIntervalIsPercent(bool on) { SB(this->vehicle_flags, VF_SERVINT_IS_PERCENT, 1, on); }
 
+	inline bool AutomaticSeparationIsEnabled() const { return (this->orders.list == nullptr) ? false : this->orders.list->AutomaticSeparationIsEnabled(); }
+	inline void SetAutomaticSeparationIsEnabled(bool enabled) const { if (this->orders.list != nullptr) this->orders.list->SetAutomaticSeparationIsEnabled(enabled); }
```
bit more nitpicking, but I would keep the style as above, and a newline extra between `inline` commands.

twpol replied:
Fixed

TrueBrain reviewed (changes requested):
Did not test it yet; will do after you rebase. Some minor coding style thingies, but overall it looks really good :D Happy to see such a clean implementation of such feature :) Awesome work, tnx!

twpol commented:
> Did not test it yet; will do after you rebase. Some minor coding style thingies, but overall it looks really good :D Happy to see such a clean implementation of such feature :) Awesome work, tnx!

Thanks!

Rebased onto latest code and fixed everything I can, further suggestions welcome.

andythenorth commented:
I saw this deadlock, with all separated vehicles waiting in stations.  Using 'stop all vehicles' and 'start all vehicles' (for each vehicle type) cleared it.

Happened after loading a savegame, but that might be coincidence.  

I can't repro by reloading the save.  The savegame also uses unreleased grfs and grfs have been reloaded many times, so it's not helpful for debugging.

Probably not much you can do with this report, but if I see it again, I'll try and get a repro.

twpol commented:
> Rebased onto latest code and fixed everything I can, further suggestions welcome.

@TrueBrain Are further changes needed? I've commented on and updated everything I can, so I am not sure what is expected of me next, if not for another review pass.

TrueBrain commented:
No, you are good for now :) We now have to review the code, and talk over what we think about a feature like this. Sorry, the removing of "reviewer" is just as I don't want others to think I am going to do it; I might, but it can also be another dev'er :P

If you have any idea why the problem that @andythenorth happened, that would be great too.

twpol commented:
Okay, great, thanks!

I'll see if I can figure out what happened with https://github.com/OpenTTD/OpenTTD/pull/8342#issuecomment-748331405

twpol commented:
I haven't been able to figure out a possible cause of the deadlock in https://github.com/OpenTTD/OpenTTD/pull/8342#issuecomment-748331405 :(

I've made some small improvements to the code and am going to rebase onto the latest code now.

andythenorth commented:
> I haven't been able to figure out a possible cause of the deadlock in [#8342 (comment)]

Some things don't get a reliable repro unless they get enough players using them.  The deadlocks were trivially resolvable by using stop-all, start-all for each vehicle type.  

My vote would be (if this passes review) to push it to a nightly and let players find out if the deadlock recurs.  That has the risk we have to back it out, causing disappointment, but eh, fair gamble.  Also...not my decision anyway :)

JGRennison commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
+	/* Check this feature is enabled on the vehicle's orders */
+	if (!this->AutomaticSeparationIsEnabled()) return;
+
+	/* Only perform the separation at the first manual order (saves on storage) */
```
The first manual order is not necessarily a suitable point to wait.

In particular it would be advisable to also exclude waypoint orders, conditional orders and depot orders.

twpol replied:
I have added to the check the existing `o->CanLoadOrUnload()` which looks like it will exclude all undesirable orders (when combined with the `OT_IMPLICIT` check).

JGRennison commented on src/vehicle.cpp:
```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
+		if (!(v->vehstatus & VS_STOPPED)) {
+			vehicles++;
+			/* Count vehicles queing for the first manual order but not currently in the station */
+			if (v != this && v->cur_speed == 0 && v->cur_implicit_order_index == first_manual_order && !v->current_order.IsType(OT_LOADING)) {
```
What's the rationale for this heuristic?

That a vehicle is stopped somewhere does not necessarily imply that it is queuing immediately before the target station.
If a train is quite normally stopped at a red signal en-route it would have a disproportionate effect on the separation.

twpol replied:
This heuristic is the closest I could get to the desired information. If you can do better, please suggest how. :)

The comment below this code should explain why we need to count queuing vehicles, and, since the status is only checked when a vehicle becomes ready to depart, in practise it does not seem problematic to have a few (more/fewer) vehicles miscategorised here.

JGRennison replied:
I can see where you're coming from. Something like this probably needs play-testing on a variety of network types to see how it performs.

Given that separation is only applied at the first manual order, that order seems likely to be intended to be a terminus of some kind. The comments imply that you're thinking of a single platform type layout with a queue to get in. This would seem like an odd design choice at the player level.
Any congestion on the approach to the first manual order would be liable to make vehicles leave the terminus early. This could disproportionately impact the simple case where the order list has only two stations in it, for example.

2TallTyler replied:
I suggested this elsewhere, but the ideal place to separate vehicles is a depot. I don't think it would be unreasonable to only separate at the first depot encountered by a train, with the Autoseparate button giving the user a command error if the order list does not include a depot. That would avoid the primary complaint/concern I've heard about  autoseparation, which is clogs due to vehicles attempting to space themselves out.

JGRennison replied:
Using a depot like this is a bit unintuitive if the cargo flow is bidirectional (i.e passengers, mail).
You don't really want your passengers/mail ageing in a depot waiting for a departure slot to be able to unload.
You could use an unload order, then a depot order, then a load order, but users would reasonably ask why they can't just use a terminus with enough platforms instead.

2TallTyler replied:
Or let CargoDist handle it automatically, but yes, users might complain. That said, it's still the most durable way to handle waiting time.

twpol replied:
If we don't think the queuing applies to other types of vehicles, I am happy to condition it, but it is required for road vehicles from playtesting.

JGRennison commented:
You might want to consider disabling separation if conditional orders are present, as the RTT estimation may produce nonsensical results.

JGRennison commented on src/order_cmd.cpp:
```diff
@@ -1517,6 +1517,40 @@ static bool CheckAircraftOrderDistance(const Aircraft *v_new, const Vehicle *v_o
+	if (ret.Failed()) return ret;
+
+	if (flags & DC_EXEC) {
+		vehicle->SetAutomaticSeparationIsEnabled(p2);
```
If automatic separation was previously enabled, this should probably call ResetAutomaticSeparation, as otherwise any old values left in `first_order_last_departure` will take effect and cause bugs.

twpol replied:
This code resets the data when you _disable_ automatic separation, and elsewhere we reset for every case which could cause issues with the data (e.g. when a vehicle is manually stopped).

I am happy to flip this to reset when you _enable_ automatic separation, or even in both directions, but I don't currently see how it would make any difference (outside of bugs elsewhere, which should be fixed in their own right).

Please let me know what I've missed here.

twpol committed:
Cleanup: Ensure that separation happens at station with load/unload

twpol committed:
Cleanup: Reset separation when skipping orders

2TallTyler commented on src/order_cmd.cpp:
```diff
@@ -1466,6 +1467,36 @@ static bool CheckAircraftOrderDistance(const Aircraft *v_new, const Vehicle *v_o
+/**
+ * Enable or disable automatic separation for a vehicle's order list
+ * @param flags operation to perform
+ * @param veh vehicle who's order list is being modified
```
`who's` -> `whose`

2TallTyler commented on src/saveload/saveload.h:
```diff
@@ -331,6 +331,7 @@ enum SaveLoadVersion : uint16_t {
 	SLV_CUSTOM_SUBSIDY_DURATION,            ///< 292  PR#9081 Configurable subsidy duration.
 	SLV_SAVELOAD_LIST_LENGTH,               ///< 293  PR#9374 Consistency in list length with SL_STRUCT / SL_STRUCTLIST / SL_DEQUE / SL_REFLIST.
 	SLV_RIFF_TO_ARRAY,                      ///< 294  PR#9375 Changed many CH_RIFF chunks to CH_ARRAY chunks.
+	SLV_AUTOMATIC_SEPARATION,               ///< 295  PR#8342 Allow automatically separating vehicles in shared orders.
```
This is rebased incorrectly. It needs to be the last entry in the list (324 as I write this), just above SL_MAX_VERSION.

twpol replied:
Whoops, not sure how I missed this. Fixed. :)

2TallTyler commented on src/vehicle.cpp:
```diff
@@ -2360,6 +2365,96 @@ void Vehicle::HandleLoading(bool mode)
+ * it is ready to depart)
+ */
+bool Vehicle::IsWaitingForAutomaticSeparation() const {
+	TimerGameTick::Ticks now = TimerGameCalendar::date.base() * Ticks::DAY_TICKS + TimerGameCalendar::date_fract;
```
This calculation works because of order of operation, but would be nicer with parentheses. 🙂

2TallTyler commented on src/vehicle.cpp:
```diff
@@ -2360,6 +2365,96 @@ void Vehicle::HandleLoading(bool mode)
+	/* Check this feature is enabled on the vehicle's orders */
+	if (!this->AutomaticSeparationIsEnabled()) return;
+
+	/* Only perform the separation at the first manual order which can load/unload (saves on storage) */
```
This is a really good feature for network planning, because the player can make the first order a large station where vehicles can wait out the correction period. 

It would be even better if (as a separate PR), you allowed vehicles to wait in depots. JGRPP has an implementation of this that you could look at. Then players could set a depot as the first order and avoid delaying other vehicles. 😃

2TallTyler commented on src/vehicle.cpp:
```diff
@@ -2360,6 +2365,96 @@ void Vehicle::HandleLoading(bool mode)
+		/* A stopped vehicle is not included; it might be stopped by player or parked in a depot */
+		if (!(v->vehstatus & VS_STOPPED)) {
+			vehicles++;
+			/* Count vehicles queing for the first manual order but not currently in the station */
```
`queing` -> `queuing`

2TallTyler commented on src/vehicle.cpp:
```diff
@@ -2360,6 +2365,96 @@ void Vehicle::HandleLoading(bool mode)
+		if (!(v->vehstatus & VS_STOPPED)) {
+			vehicles++;
+			/* Count vehicles queing for the first manual order but not currently in the station */
+			if (v != this && v->cur_speed == 0 && v->cur_implicit_order_index == first_manual_order && !v->current_order.IsType(OT_LOADING)) {
```
Does this count vehicles heading for the first order but not moving due to being stopped at a signal?

twpol replied:
Yes. It will count all stationary vehicles heading to their first stop, except for those loading or which have been manually stopped.

2TallTyler commented on src/lang/english.txt:
```diff
@@ -4485,6 +4485,9 @@ STR_ORDERS_DELETE_ALL_TOOLTIP                                   :{BLACK}Delete a
 STR_ORDERS_STOP_SHARING_TOOLTIP                                 :{BLACK}Stop sharing the order list. Ctrl+Click additionally deletes all orders for this vehicle
 
+STR_ORDERS_AUTOMATIC_SEPARATION                                 :{BLACK}Automatic separation
+STR_ORDERS_AUTOMATIC_SEPARATION_TOOLTIP                         :{BLACK}Automatically separate all vehicles sharing this order
```
`this order` -> `these orders` ?

twpol committed:
Cleanup: Update stored date when it is shifted

twpol committed:
Cleanup: Fix typos and other review corrections

twpol committed:
Cleanup: Convert last departure date to tick counter

twpol committed:
Cleanup: Fix bad debug logging with TickCounter

twpol committed:
Cleanup: Apply automatic separation to fully timetabled vehicles
